be/src/exec/pipeline/fragment_context.cpp

```diff
@@ -180,6 +180,7 @@ void FragmentContext::set_stream_load_contexts(const std::vector<StreamLoadConte
 void FragmentContext::cancel(const Status& status) {
     if (!status.ok() && _runtime_state != nullptr && _runtime_state->query_ctx() != nullptr) {
         _runtime_state->query_ctx()->release_workgroup_token_once();
+        _driver_token.reset();
```

Comment:
would it be reset more than once?

Reply:
For now, it ensures that FE only cancel each fragment once. However, to avoid potential bug in the future, add new method `release_driver_token_once`.

Reply:
Oh,  just move `_driver_token.reset` to `set_final_status`, since `FragmentContext::cancel` calls `FragmentContext::cancel`, which already apply CAS.

```cpp
void FragmentContext::set_final_status(const Status& status) {
    if (_final_status.load() != nullptr) {
        return;
    }
    Status* old_status = nullptr;
    if (_final_status.compare_exchange_strong(old_status, &_s_status)) {
        _s_status = status;

        _driver_token.reset();
```